src/Generators/Microsoft.Gen.Logging/Parsing/Parser.cs

```diff
@@ -639,67 +617,125 @@ private void CheckTagNamesAreUnique(LoggingMethod lm, Dictionary<LoggingMethodPa
+        //
+        //     [LoggerMessage(EventId = 0, Level = LogLevel.Debug, Message = ""M1"")]
+        //     public partial void M1(); // The ILogger primary constructor parameter cannot be used here.
+        // }
+#pragma warning restore S125 // Sections of code should not be commented out
+
+        HashSet<string> shadowedNames = new(StringComparer.Ordinal);
+
+        while (currentClassType is { SpecialType: not SpecialType.System_Object })
         {
-            foreach (var m in type.GetMembers())
+            foreach (ISymbol ms in currentClassType.GetMembers())
             {
-                if (m is IPropertySymbol p)
+                // we support both fields and properties
+                ITypeSymbol? typeSymbol = ms switch
                 {
-                    var gm = p.GetMethod;
-                    if (gm != null)
-                    {
-                        if (ParserUtilities.IsBaseOrIdentity(gm.ReturnType, symbol, _compilation))
-                        {
-                            if (!sm.Compilation.IsSymbolAccessibleWithin(gm, originType, type))
-                            {
-                                continue;
-                            }
+                    IFieldSymbol fs => fs.Type,
+                    IPropertySymbol ps => ps.Type,
+                    _ => default,
+                };
 
-                            if (member == null)
-                            {
-                                member = p.Name;
-                                isNullable = gm.ReturnType.NullableAnnotation == NullableAnnotation.Annotated;
-                            }
-                            else
-                            {
-                                return (null, p, isNullable);
-                            }
-                        }
+                if (typeSymbol is null)
+                {
+                    continue;
+                }
+
+                if (!onMostDerivedType && ms.DeclaredAccessibility == Accessibility.Private)
+                {
+                    continue;
+                }
+
+                if (!ms.CanBeReferencedByName)
+                {
+                    continue;
+                }
+
+                if (ParserUtilities.IsBaseOrIdentity(typeSymbol, loggerSymbol, _compilation))
+                {
+                    if (loggerMember == null)
+                    {
+                        loggerMember = ms.Name;
+                        isNullable = typeSymbol.NullableAnnotation == NullableAnnotation.Annotated;
+                    }
+                    else if (!onMostDerivedType)
+                    {
+                        // we found a public logger member on a base class, prefer the existing one
+                        // because it's more likely to be the one the user intended
+                        // the dotnet logging generator doesn't support ILogger properties so this matches that behavior
```

Comment:
```suggestion
                        // We found a public logger member on a base class, prefer the existing one
                        // because it's more likely to be the one the user intended.
                        // The .NET runtime logging generator doesn't support ILogger properties, so this matches that behavior.
```